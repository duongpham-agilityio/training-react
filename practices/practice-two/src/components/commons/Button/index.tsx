import { ButtonHTMLAttributes, memo } from 'react';

// Styles
import buttonStyles from 'components/commons/Button/index.module.css';

type Variant = 'default' | 'primary' | 'secondary' | 'danger';
type Size = 'small' | 'medium' | 'large';
type Border = 'b-md' | 'b-lg';
type Width = 'w-sm' | 'w-lg';
export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  label?: string;
  leftIcon?: string;
  rightIcon?: string;
  variant?: Variant;
  border?: Border;
  size?: Size;
  width?: Width;
};

const Icon = ({
  icon,
  subTitle = 'icon action',
  ...rest
}: {
  icon?: string;
  subTitle?: string;
}) =>
  icon ? (
    <img
      src={icon}
      alt={subTitle}
      className={buttonStyles.icon}
      {...rest}
      width="22px"
      height="25px"
    />
  ) : null;

const ButtonIcon = memo(Icon);

const Button = (props: ButtonProps) => {
  const {
    label = '',
    leftIcon = '',
    rightIcon = '',
    className = '',
    variant = 'default',
    size = 'medium',
    border = 'b-md',
    width = 'w-sm',
    type = 'button',
    ...rest
  } = props;

  const classes = `${buttonStyles.btn} ${buttonStyles[variant]}
  ${buttonStyles[size]} ${buttonStyles[border]}
  ${buttonStyles[width]} ${className}`;

  return (
    <button className={classes} type={type} {...rest}>
      <ButtonIcon
        icon={leftIcon}
        subTitle="This is the left icon for the button"
      />

      {label}

      <ButtonIcon
        icon={rightIcon}
        subTitle="This is the right icon for the button"
      />
    </button>
  );
};

export default memo(Button);
