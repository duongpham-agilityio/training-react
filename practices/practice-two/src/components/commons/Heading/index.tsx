import { HTMLAttributes, createElement, memo } from 'react';

// Styles
import stylesHeading from 'components/commons/Heading/index.module.css';

type Size = 'sm' | 'md' | 'lg' | 'xl';
type TagName = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
export type HeadingProps = HTMLAttributes<HTMLHeadingElement> & {
  label?: string;
  tag?: TagName;
  size?: Size;
};

const Heading = (props: HeadingProps) => {
  const {
    children,
    className,
    label = '',
    size = 'sm',
    tag = 'h2',
    ...rest
  } = props;
  const Component = createElement(
    tag,
    {
      ...rest,
      className: `${stylesHeading.heading} ${stylesHeading[size]} ${
        className ?? ''
      }`,
    },
    children || label
  );

  return Component;
};

export default memo(Heading);
