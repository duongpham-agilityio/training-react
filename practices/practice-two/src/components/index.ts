export {
  default as Notification,
  type NotificationProps,
} from './Notification';

export { default as Popup, type PopupProps } from './Popup';
export { default as Card, type CardProps } from './Card';
export { default as CardSkeleton } from './Card/Skeleton';
export { default as Error } from './Error';
