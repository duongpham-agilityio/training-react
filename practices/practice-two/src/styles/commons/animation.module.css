@keyframes skeleton {
  from {
    left: -30%;
  }

  to {
    left: 120%;
  }
}
