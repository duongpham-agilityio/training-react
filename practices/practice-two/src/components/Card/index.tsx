import { MouseEvent, memo } from 'react';
import { Link, LinkProps } from 'react-router-dom';

// Component
import { Button, Heading } from 'components/commons';

// Styles
import styles from 'components/Card/index.module.css';

// Assets
import NextIcon from 'assets/icons/next.svg';
import TrashIcon from 'assets/icons/trash.svg';

export type CardProps = Omit<LinkProps, 'to'> & {
  href?: string;
  title: string;
  description: string;
  publishedDate: string;
  imageUrl: string;
  onDeleteBook?: (_event: MouseEvent) => void;
};

const Card = (props: CardProps) => {
  const {
    title,
    description,
    publishedDate,
    imageUrl,
    href = '',
    onDeleteBook,
  } = props;

  return (
    <Link to={href} className={styles.card}>
      <div className={styles.cardItem}>
        <Heading label={title} className={styles.heading} />
        <p className={styles.description}>{description}</p>
        <div className={styles.action}>
          <Button leftIcon={NextIcon} variant="primary" />
          <Button
            leftIcon={TrashIcon}
            variant="danger"
            onClick={onDeleteBook}
          />
        </div>
      </div>
      <div className={styles.cardItem}>
        <p className={styles.time}>{publishedDate}</p>
        {imageUrl ? (
          <img src={imageUrl} alt={title} className={styles.image} />
        ) : (
          <div className={`${styles.image} ${styles.fake}`}></div>
        )}
      </div>
    </Link>
  );
};

export default memo(Card);
