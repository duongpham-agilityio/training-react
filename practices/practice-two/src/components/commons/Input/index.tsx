import {
  InputHTMLAttributes,
  MouseEvent,
  Ref,
  TextareaHTMLAttributes,
  forwardRef,
  memo,
} from 'react';

// Components
import Button from '../Button';

// Style
import inputStyles from 'components/commons/Input/index.module.css';
import textAreaStyle from 'components/commons/Input/textArea.module.css';

export type InputProps = InputHTMLAttributes<HTMLInputElement> &
  TextareaHTMLAttributes<HTMLTextAreaElement> & {
    variant?: 'input' | 'area';
    leftIcon?: string;
    rightIcon?: string;
    customRef?: Ref<HTMLInputElement | HTMLTextAreaElement>;
    leftIconClick?: (_event: MouseEvent) => void;
    rightIconClick?: (_event: MouseEvent) => void;
  };

const TextArea = (props: Omit<InputProps, 'variant'>) => {
  const { value = '', className = '', ...rest } = props;

  return (
    <textarea
      className={`${textAreaStyle.text} ${className}`}
      value={value}
      {...rest}
    ></textarea>
  );
};

const Input = (props: Omit<InputProps, 'variant'>) => {
  const {
    leftIcon = '',
    rightIcon = '',
    customRef,
    className = '',
    type = 'text',
    placeholder = 'Enter text...',
    leftIconClick,
    rightIconClick,
    ...rest
  } = props;

  return (
    <div className={`${inputStyles.input} ${className}`}>
      <Button
        leftIcon={leftIcon}
        style={{
          backgroundColor: 'transparent',
          padding: '0',
          width: 'fit-content',
          height: 'fit-content',
        }}
        onClick={leftIconClick}
        aria-label="Search button"
      />

      <input
        className={inputStyles.value}
        type={type}
        placeholder={placeholder}
        ref={customRef as Ref<HTMLInputElement>}
        {...rest}
      />

      <Button
        leftIcon={rightIcon}
        style={{
          backgroundColor: 'transparent',
          padding: '0',
          width: 'fit-content',
          height: 'fit-content',
        }}
        onClick={rightIconClick}
        aria-label="The right button"
      />
    </div>
  );
};

const Component = (
  { variant, ...rest }: Omit<InputProps, 'customRef'>,
  ref: Ref<HTMLInputElement | HTMLTextAreaElement>
) => {
  const component = {
    input: Input,
    area: TextArea,
  };

  const Render = component[variant ?? 'input'];

  return <Render {...rest} customRef={ref} />;
};

export default memo(forwardRef(Component));
