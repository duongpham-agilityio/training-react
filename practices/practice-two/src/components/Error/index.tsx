// Component
import { Heading } from 'components/commons';

// Constants
import { MESSAGES } from '@constants';

// Styles
import styles from 'components/Error/index.module.css';

export interface ErrorProps {
  title?: string;
  description?: string;
}

const Error = ({
  title = MESSAGES.ERROR_TITLE,
  description = MESSAGES.ERROR_DESCRIPTION,
}: ErrorProps) => (
  <section className={styles.empty}>
    <Heading label={title} className={styles.emptyTitle} size="sm" />
    <p className={styles.emptyDescription}>{description}</p>
  </section>
);

export default Error;
