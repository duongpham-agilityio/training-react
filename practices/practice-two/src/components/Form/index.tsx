import { MouseEvent, memo, useCallback } from 'react';
import isEqual from 'react-fast-compare';

// Hooks
import { useBookForm, FormData } from 'hooks';

// Components
import { Button, Heading, Input } from 'components/commons';
import BookRecommend from './Recommend';
import ChooseImage from './ChooseImage';

// Styles
import styles from 'components/Form/index.module.css';

export type FormProps = {
  value: FormData;
  title?: string;
  className?: string;
  type?: 'update' | 'create';
  onClose?: () => void;
  onSubmit: (data: FormData) => Promise<void>;
};

const Form = (props: FormProps) => {
  const {
    value: data,
    title = 'Create a new book marker',
    className = '',
    type = 'create',
    onClose,
    onSubmit,
  } = props;
  const {
    value: { author, description, imageURL, name, publishDate, imageName },
    refImage,
    isUpload,
    isSubmit,
    booksRecommended,
    handleSelectRecommended,
    resetRecommended,
    onChange,
    onSubmit: onSubmitOnHook,
  } = useBookForm(data, type, onSubmit);

  const chooseImage = useCallback(() => {
    refImage.current?.click();
  }, [refImage]);

  const resetRecommendedData = useCallback(
    (e: MouseEvent) => {
      e.stopPropagation();

      resetRecommended([]);
    },
    [resetRecommended]
  );

  return (
    <section className={styles.overlay} onClick={onClose}>
      <form
        className={`${styles.form} ${className}`}
        action="#"
        method="POST"
        onSubmit={onSubmitOnHook}
        onClick={resetRecommendedData}
        data-testid="book-form"
      >
        <Heading label={title} className={styles.heading} />

        <section className={styles.content}>
          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Book name
              </label>
              <Input
                value={name}
                placeholder="Book name"
                name="name"
                id="bookName"
                onChange={onChange}
              />

              {!!booksRecommended.length && (
                <BookRecommend
                  books={booksRecommended}
                  onSelect={handleSelectRecommended}
                />
              )}
            </div>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="author">
                Author
              </label>
              <Input
                value={author}
                placeholder="Author"
                name="author"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="publishedDate">
                Published date
              </label>
              <Input
                type="date"
                value={publishDate}
                name="publishDate"
                onChange={onChange}
              />
            </div>
            <div className={`${styles.formField} ${styles.upload}`}>
              <div className={styles.uploadItem}>
                <label className={styles.label} htmlFor="imageURL">
                  Upload
                </label>
                <Input
                  ref={refImage}
                  hidden
                  className={styles.input}
                  type="file"
                  name="imageURL"
                  id="imageURL"
                  accept=".png,.jpeg"
                  onChange={onChange}
                />

                <ChooseImage name={imageName} chooseHandler={chooseImage} />
              </div>
              <div className={styles.uploadItem}>
                {imageURL && (
                  <img
                    src={imageURL}
                    alt="image book"
                    className={styles.image}
                  />
                )}
              </div>
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Description
              </label>
              <Input
                variant="area"
                className={styles.text}
                value={description}
                placeholder="Book name"
                name="description"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.action}>
            <Button
              label="Cancel"
              width="w-lg"
              border="b-lg"
              onClick={onClose}
              disabled={isSubmit}
            />
            <Button
              disabled={isUpload || isSubmit}
              type="submit"
              label="Save"
              variant={isUpload ? 'default' : 'primary'}
              width="w-lg"
              border="b-lg"
            />
          </div>
        </section>
      </form>
    </section>
  );
};

export default memo(Form, isEqual);
