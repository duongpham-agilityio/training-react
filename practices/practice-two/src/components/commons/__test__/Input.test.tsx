import { fireEvent, render } from '@testing-library/react';

import { Input } from 'components/commons';

const setup = (callback: () => void) =>
  render(<Input value="" onChange={callback} placeholder="Enter" />);

describe('Input', () => {
  it('match snapshot', () => {
    const { asFragment } = setup(jest.fn());

    expect(asFragment()).toMatchSnapshot();
  });

  it('render', () => {
    const { getByPlaceholderText } = setup(jest.fn());

    expect(getByPlaceholderText(/Enter/i)).toBeInTheDocument();
  });

  it('action', () => {
    const mockOnChange = jest.fn();
    const { getByPlaceholderText } = setup(mockOnChange);
    const input = getByPlaceholderText(/Enter/i);

    fireEvent.change(input, {
      target: {
        value: '1',
      },
    });

    expect(mockOnChange).toHaveBeenCalled();
  });

  it('Render with textArea', () => {
    const { getByTestId } = render(
      <Input variant="area" data-testid="custom-element" />
    );

    const textArea = getByTestId('custom-element');

    expect(textArea).toBeInTheDocument();
  });
});
