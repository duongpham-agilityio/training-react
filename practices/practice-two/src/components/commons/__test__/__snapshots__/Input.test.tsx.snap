// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input match snapshot 1`] = `
<DocumentFragment>
  <div
    class="undefined "
  >
    <button
      aria-label="Search button"
      class="undefined undefined
  undefined undefined
  undefined "
      style="background-color: transparent; padding: 0px;"
      type="button"
    />
    <input
      placeholder="Enter"
      type="text"
      value=""
    />
    <button
      aria-label="The right button"
      class="undefined undefined
  undefined undefined
  undefined "
      style="background-color: transparent; padding: 0px;"
      type="button"
    />
  </div>
</DocumentFragment>
`;
